### ISSUE TITLE
Failed to upload terminology with hapi-fhir-cli v4.0.0-SNAPSHOT

### ISSUE THREAD

**Issue by Bugerman58** [jamesagnew/hapi-fhir#1394]
**Describe the bug**
HAPI server fails to upload latest LOINC, and raises exception: `Failed to call access method: java.lang.IllegalArgumentException: theValue must not be null or empty`

**To Reproduce**
1. Clone project
1. Change directory to ./hapi-fhir/hapi-fhir-cli
2. Compile tool: mvn clean package
3. Change directory to ./hapi-fhir-cli-app/target
4. Run server: `java -jar hapi-fhir-cli.jar  run-server -v r4`
```shell
------------------------------------------------------------
🔥  HAPI FHIR 4.0.0-SNAPSHOT - Command Line Tool
------------------------------------------------------------
Process ID                      : 19980@Mac-mini-Anton.local
Max configured JVM memory (Xmx) : 2.0GB
Detected Java version           : 12.0.1
------------------------------------------------------------
2019-07-25 13:29:19.492 [main] INFO  c.u.f.c.RunServerCommand Preparing HAPI FHIR JPA server on port 8080
2019-07-25 13:29:19.771 [main] INFO  c.u.f.c.RunServerCommand Starting HAPI FHIR JPA server in R4 mode
2019-07-25 13:29:34.21 [scheduledExecutorService-5] WARN  o.h.e.j.s.SqlExceptionHelper SQL Warning Code: 10000, SQLState: 01J01
2019-07-25 13:29:34.24 [scheduledExecutorService-5] WARN  o.h.e.j.s.SqlExceptionHelper Database 'directory:target/jpaserver_derby_files' not created, connection made to existing database instead.
2019-07-25 13:29:34.26 [scheduledExecutorService-4] WARN  o.h.e.j.s.SqlExceptionHelper SQL Warning Code: 10000, SQLState: 01J01
2019-07-25 13:29:34.26 [scheduledExecutorService-4] WARN  o.h.e.j.s.SqlExceptionHelper Database 'directory:target/jpaserver_derby_files' not created, connection made to existing database instead.
2019-07-25 13:29:34.30 [scheduledExecutorService-3] WARN  o.h.e.j.s.SqlExceptionHelper SQL Warning Code: 10000, SQLState: 01J01
2019-07-25 13:29:34.30 [scheduledExecutorService-3] WARN  o.h.e.j.s.SqlExceptionHelper Database 'directory:target/jpaserver_derby_files' not created, connection made to existing database instead.
2019-07-25 13:29:34.60 [scheduledExecutorService-2] WARN  o.h.e.j.s.SqlExceptionHelper SQL Warning Code: 10000, SQLState: 01J01
2019-07-25 13:29:34.60 [scheduledExecutorService-2] WARN  o.h.e.j.s.SqlExceptionHelper Database 'directory:target/jpaserver_derby_files' not created, connection made to existing database instead.
2019-07-25 13:29:34.380 [main] INFO  c.u.f.c.RunServerCommand Server started on port 8080
2019-07-25 13:29:34.380 [main] INFO  c.u.f.c.RunServerCommand Web Testing UI : http://localhost:8080/
2019-07-25 13:29:34.380 [main] INFO  c.u.f.c.RunServerCommand Server Base URL: http://localhost:8080/baseR4/
```
5. In another tab run terminology upload: `java -jar hapi-fhir-cli.jar upload-terminology -v r4 -u http://loinc.org -t http://localhost:8080/baseR4 -d ~/Downloads/Loinc_2.66.zip`
```shell
------------------------------------------------------------
🔥  HAPI FHIR 4.0.0-SNAPSHOT - Command Line Tool
------------------------------------------------------------
Process ID                      : 20086@Mac-mini-Anton.local
Max configured JVM memory (Xmx) : 2.0GB
Detected Java version           : 12.0.1
------------------------------------------------------------
2019-07-25 13:50:46.733 [main] INFO  c.u.f.c.UploadTerminologyCommand Beginning upload - This may take a while...
2019-07-25 13:50:52.537 [main] ERROR ca.uhn.fhir.cli.App Error during execution: 
ca.uhn.fhir.rest.server.exceptions.InternalErrorException: HTTP 500 Server Error: Failed to call access method: java.lang.IllegalArgumentException: theValue must not be null or empty
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
	at ca.uhn.fhir.rest.server.exceptions.BaseServerResponseException.newInstance(BaseServerResponseException.java:302)
	at ca.uhn.fhir.rest.client.impl.BaseClient.invokeClient(BaseClient.java:351)
	at ca.uhn.fhir.rest.client.impl.GenericClient$BaseClientExecutable.invoke(GenericClient.java:431)
	at ca.uhn.fhir.rest.client.impl.GenericClient$OperationInternal.execute(GenericClient.java:1170)
	at ca.uhn.fhir.cli.UploadTerminologyCommand.run(UploadTerminologyCommand.java:109)
	at ca.uhn.fhir.cli.BaseApp.run(BaseApp.java:251)
	at ca.uhn.fhir.cli.App.main(App.java:43)
2019-07-25 13:50:52.538 [Thread-0] INFO  ca.uhn.fhir.cli.App HAPI FHIR is shutting down...
```
6. Server stacktrace:
```shell
2019-07-25 13:50:51.00 [qtp875016237-16] ERROR c.u.f.r.s.i.ExceptionHandlingInterceptor Failure during REST processing
ca.uhn.fhir.rest.server.exceptions.InternalErrorException: Failed to call access method: java.lang.IllegalArgumentException: theValue must not be null or empty
	at ca.uhn.fhir.rest.server.method.BaseMethodBinding.invokeServerMethod(BaseMethodBinding.java:244)
	at ca.uhn.fhir.rest.server.method.OperationMethodBinding.invokeServer(OperationMethodBinding.java:312)
	at ca.uhn.fhir.rest.server.method.BaseResourceReturningMethodBinding.doInvokeServer(BaseResourceReturningMethodBinding.java:243)
	at ca.uhn.fhir.rest.server.method.BaseResourceReturningMethodBinding.invokeServer(BaseResourceReturningMethodBinding.java:380)
	at ca.uhn.fhir.rest.server.method.OperationMethodBinding.invokeServer(OperationMethodBinding.java:287)
	at ca.uhn.fhir.rest.server.RestfulServer.handleRequest(RestfulServer.java:998)
	at ca.uhn.fhir.rest.server.RestfulServer.doPost(RestfulServer.java:346)
	at ca.uhn.fhir.rest.server.RestfulServer.service(RestfulServer.java:1656)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:867)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:542)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1588)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1557)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:502)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:364)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
	at java.base/java.lang.Thread.run(Thread.java:835)
Caused by: java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at ca.uhn.fhir.rest.server.method.BaseMethodBinding.invokeServerMethod(BaseMethodBinding.java:239)
	... 38 common frames omitted
Caused by: java.lang.IllegalArgumentException: theValue must not be null or empty
	at ca.uhn.fhir.util.ValidateUtil.isNotBlankOrThrowIllegalArgument(ValidateUtil.java:50)
	at ca.uhn.fhir.jpa.entity.TermConceptDesignation.setValue(TermConceptDesignation.java:115)
	at ca.uhn.fhir.jpa.term.loinc.LoincHandler.accept(LoincHandler.java:71)
	at ca.uhn.fhir.jpa.term.TerminologyLoaderSvcImpl.iterateOverZipFile(TerminologyLoaderSvcImpl.java:171)
	at ca.uhn.fhir.jpa.term.TerminologyLoaderSvcImpl.processLoincFiles(TerminologyLoaderSvcImpl.java:424)
	at ca.uhn.fhir.jpa.term.TerminologyLoaderSvcImpl.loadLoinc(TerminologyLoaderSvcImpl.java:240)
	at ca.uhn.fhir.jpa.provider.BaseTerminologyUploaderProvider.handleUploadExternalCodeSystem(BaseTerminologyUploaderProvider.java:123)
	at ca.uhn.fhir.jpa.provider.r4.TerminologyUploaderProviderR4.uploadExternalCodeSystem(TerminologyUploaderProviderR4.java:48)
	... 43 common frames omitted
```

**Expected behavior**
Successful upload expected.

**Environment:**
 - HAPI FHIR Version: 4.0.0-SNAPSHOT
 - OS: macOS v10.14.6
 - Java version: 12.0.1
 - LOINC version: 2.66

**Comment by jamesagnew** [comment]
@dmuylwyk would you mind to give this a look?

**Comment by dmuylwyk** [comment]
@jamesagnew, sure thing.

**Comment by dmuylwyk** [comment]
Hi, @Bugerman58. [`TRM_CONCEPT_DESIG.VAL`](https://github.com/jamesagnew/hapi-fhir/blob/master/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/TermConceptDesignation.java#L57) is not nullable.

I suspect we simply need to ignore rows where no value is provided. I will attempt a fix either later today or tomorrow.

**Comment by dmuylwyk** [comment]
This just needs a null check on this line:

https://github.com/jamesagnew/hapi-fhir/blob/master/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/loinc/LoincHandler.java#L67

**Comment by dmuylwyk** [comment]
That's a null check on `shortName`.

**Comment by dmuylwyk** [comment]
Working branch: [`1394-failed-to-upload-terminology-with-hapi-fhir-cli-v4-0-0-snapshot`](https://github.com/jamesagnew/hapi-fhir/tree/1394-failed-to-upload-terminology-with-hapi-fhir-cli-v4-0-0-snapshot)

**Comment by dmuylwyk** [comment]
Pull request: https://github.com/jamesagnew/hapi-fhir/pull/1396

**Comment by dmuylwyk** [comment]
@Bugerman58, fix has been merged into `master`. Thanks!

**Comment by Bugerman58** [comment]
@dmuylwyk Thank you!

**Comment by dmuylwyk** [comment]
@Bugerman58, my pleasure.

Please leave a comment to confirm your upload succeeds. : )

**Comment by Bugerman58** [comment]
Sorry, but i have same error message coming :(

**Comment by dmuylwyk** [comment]
Darn. I suppose it's never that easy, is it?

@Bugerman58, did you pull the latest `master` and rebuild?

I will try to reproduce this locally later this morning.

**Comment by Bugerman58** [comment]
@dmuylwyk Yes, of course, I did pull project of `master` branch(got your commit) and try repeat all steps.

**Comment by dmuylwyk** [comment]
Will investigate further today.

**Comment by dmuylwyk** [comment]
Hi, @Bugerman58. I'm unable to reproduce your issue locally. Here's what I've done:

1. pull the latest `master` for HAPI FHIR
2. in the root project directory, build with `mvn clean install -Dmaven.test.skip -U`
3. create a new directory outside the project directory called `hapi-fhir-server`
4. in this new directory, put `Loinc_2.66.zip` and `loincupload.properties`

These are the contents of `loincupload.properties`:

```bash
# This is the version identifier for the AnswerList file
answerlist.version=Loinc_2.66_AnswerFile_2.66

# This is the version identifier for uploaded ConceptMap resources
conceptmap.version=Loinc_2.66
```

5. start the server:

```bash
~/i/hapi-fhir-server $ java -jar ~/intellij-idea-workspace/hapi-fhir/hapi-fhir-cli/hapi-fhir-cli-app/target/hapi-fhir-cli.jar run-server -v r4
```

6. start the upload:

```
~/i/hapi-fhir-server $ java -jar ~/intellij-idea-workspace/hapi-fhir/hapi-fhir-cli/hapi-fhir-cli-app/target/hapi-fhir-cli.jar upload-terminology -v r4 -t http://localhost:8080/baseR4 -u http://loinc.org -l -d ./Loinc_2.66.zip -d ./loincupload.properties
```

Console output:

```bash
------------------------------------------------------------
🔥  HAPI FHIR 4.0.0-SNAPSHOT - Command Line Tool
------------------------------------------------------------
Process ID                      : 6766@simpatico-mbp
Max configured JVM memory (Xmx) : 3.5GB
Detected Java version           : 1.8.0_201
------------------------------------------------------------
2019-07-29 13:52:07.479 [main] INFO  c.u.f.c.UploadTerminologyCommand Beginning upload - This may take a while...
2019-07-29 13:53:05.225 [main] INFO  c.u.f.c.UploadTerminologyCommand Upload complete!
2019-07-29 13:53:05.245 [main] INFO  c.u.f.c.UploadTerminologyCommand Response:
<Parameters xmlns="http://hl7.org/fhir">
   <parameter>
      <name value="conceptCount"/>
      <valueInteger value="212701"/>
   </parameter>
   <parameter>
      <name value="target"/>
      <valueReference>
         <reference value="CodeSystem/loinc"/>
      </valueReference>
   </parameter>
</Parameters>
2019-07-29 13:53:05.246 [main] INFO  ca.uhn.fhir.cli.App HAPI FHIR is shutting down...
~/i/hapi-fhir-server $
```

Partial server log:

```bash
2019-07-29 14:53:13.582 [scheduledExecutorService-3] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41232-6
2019-07-29 14:53:13.585 [scheduledExecutorService-3] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Saved 20 deferred ValueSet resources, have 18 remaining
2019-07-29 14:53:15.652 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41357-1
2019-07-29 14:53:15.655 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41325-8
2019-07-29 14:53:15.669 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41329-0
2019-07-29 14:53:15.673 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41443-9
2019-07-29 14:53:15.677 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41123-7
2019-07-29 14:53:15.680 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41125-2
2019-07-29 14:53:15.686 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41124-5
2019-07-29 14:53:15.691 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41126-0
2019-07-29 14:53:15.695 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41233-4
2019-07-29 14:53:16.739 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41141-9
2019-07-29 14:53:16.742 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41511-3
2019-07-29 14:53:16.746 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41234-2
2019-07-29 14:53:16.750 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41127-8
2019-07-29 14:53:16.755 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41128-6
2019-07-29 14:53:16.760 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41235-9
2019-07-29 14:53:16.764 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41236-7
2019-07-29 14:53:18.493 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41241-7
2019-07-29 14:53:18.507 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Creating ValueSet: LG41238-3
2019-07-29 14:53:18.511 [scheduledExecutorService-5] INFO  c.u.f.j.t.BaseHapiTerminologySvcImpl Saved 18 deferred ValueSet resources, have 0 remaining
```

I see that you omitted `-d ./loincupload.properties` when you ran the `upload-terminology` command. However, doing so should have resulted in the following:

```bash
------------------------------------------------------------
🔥  HAPI FHIR 4.0.0-SNAPSHOT - Command Line Tool
------------------------------------------------------------
Process ID                      : 10781@simpatico-mbp
Max configured JVM memory (Xmx) : 3.5GB
Detected Java version           : 1.8.0_201
------------------------------------------------------------
2019-07-29 14:57:40.236 [main] INFO  c.u.f.c.UploadTerminologyCommand Beginning upload - This may take a while...
2019-07-29 14:57:45.262 [main] ERROR ca.uhn.fhir.cli.App Error during execution: 
ca.uhn.fhir.rest.server.exceptions.UnprocessableEntityException: HTTP 422 Unprocessable Entity: Could not find the following mandatory files in input: [loincupload.properties]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at ca.uhn.fhir.rest.server.exceptions.BaseServerResponseException.newInstance(BaseServerResponseException.java:302)
	at ca.uhn.fhir.rest.client.impl.BaseClient.invokeClient(BaseClient.java:351)
	at ca.uhn.fhir.rest.client.impl.GenericClient$BaseClientExecutable.invoke(GenericClient.java:431)
	at ca.uhn.fhir.rest.client.impl.GenericClient$OperationInternal.execute(GenericClient.java:1170)
	at ca.uhn.fhir.cli.UploadTerminologyCommand.run(UploadTerminologyCommand.java:109)
	at ca.uhn.fhir.cli.BaseApp.run(BaseApp.java:251)
	at ca.uhn.fhir.cli.App.main(App.java:43)
2019-07-29 14:57:45.264 [Thread-0] INFO  ca.uhn.fhir.cli.App HAPI FHIR is shutting down...
~/i/hapi-fhir-server $
```

and

```bash
2019-07-29 14:57:44.24 [qtp415186196-11] WARN  c.u.f.r.s.i.ExceptionHandlingInterceptor Failure during REST processing: ca.uhn.fhir.rest.server.exceptions.UnprocessableEntityException: Could not find the following mandatory files in input: [loincupload.properties]
```

Please follow my numbered steps in sequence. If it errs out again, please send me the full console output and server log.

Thanks!

**Comment by Bugerman58** [comment]
Hi, @dmuylwyk 

I've uploaded LOINC with your instructions provided. Thanks!

However, I would like to notice 2 things:
1. It is not clear how to use loincupload.properties file which contains version info. Probably need to add more detailed instructions 
 to the project documentation;
2. That looks like HAPI-server and CLI-tools supposed work on one server, is that right? I mean I was not able to upload loinc to the remote server. I had to copy the .zip file there and execute an upload-terminology command. Maybe it makes sense to add this too.

But it works and I wish to say thank you for your help.

Best regards,
Anthony
